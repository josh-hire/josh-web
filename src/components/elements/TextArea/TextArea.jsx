/* eslint-disable sonarjs/cognitive-complexity */
import clsx from 'clsx';
import PropTypes from 'prop-types';
import { useRef } from 'react';
import styles from './TextArea.module.scss';

const TextArea = ({
  className,
  disabled,
  label,
  name,
  meta,
  icon: Icon,
  inputClassName,
  input,
  required,
  message,
  onClickIcon,
  onKeyUp,
  insetIcon: InsetIcon,
  errorAutoHeight = false,
  ...props
}) => {
  const { dirty, error, touched, submitError } = meta;
  const textAreaRef = useRef();
  const textAutoGrow = () => {
    const elm = textAreaRef?.current;
    if (!elm.value) {
      elm.style.height = '168px';
    } else {
      elm.style.height = elm.scrollHeight + 'px';
    }
    if (elm.scrollHeight > 336) {
      elm.style.height = '336px';
      elm.style.overflowY = 'scroll';
    } else {
      elm.style.overflow = 'hidden';
    }
  };

  return (
    <div
      className={clsx(
        className,
        styles.container,
        Icon && styles['container-icon'],
        InsetIcon && styles['container-inset-icon'],
      )}
      {...props}>
      <label className={clsx(styles.label, disabled && styles.disabledLabel)}>
        {label}
        {required && <span className={styles.required}>*</span>}
      </label>
      <div className={styles.icon}>
        {Icon && (
          <span className={styles.icon} onClick={onClickIcon}>
            {Icon}
          </span>
        )}
      </div>
      <div className={styles['input-box']}>
        <textarea
          autoComplete="off"
          rows={'auto'}
          style={{ height: '168px' }}
          {...input}
          className={clsx(
            inputClassName,
            styles.input,
            !!input.value && styles.empty,
            (!!error || !!submitError) && (dirty || touched) && styles.errorInput,
          )}
          id={name}
          {...props}
          onInput={textAutoGrow}
          onKeyUp={onKeyUp}
          ref={textAreaRef}
        />
      </div>
      <div className={clsx(styles.message, errorAutoHeight && styles.autoHeight)}>
        {message && (
          <span className={styles.message}>
            <a>{message}</a>
          </span>
        )}
        {(!!error || !!submitError) && (dirty || touched) && (
          <small className={styles.error}>{error || submitError}</small>
        )}
      </div>
    </div>
  );
};

TextArea.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  errorAutoHeight: PropTypes.bool,
  icon: PropTypes.node,
  input: PropTypes.node,
  inputClassName: PropTypes.string,
  insetIcon: PropTypes.node,
  label: PropTypes.string,
  message: PropTypes.string,
  meta: PropTypes.object,
  name: PropTypes.string,
  onClickIcon: PropTypes.func,
  onKeyUp: PropTypes.func,
  required: PropTypes.bool,
};

TextArea.defaultProps = {
  className: '',
  disabled: false,
  errorAutoHeight: false,
  icon: null,
  input: {},
  inputClassName: '',
  insetIcon: null,
  label: '',
  message: '',
  meta: {},
  name: '',
  onClickIcon: () => {},
  onKeyUp: () => {},
  required: '',
};

export default TextArea;
