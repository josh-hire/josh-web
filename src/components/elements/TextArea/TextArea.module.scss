  .container {
    flex: flex;
  }

  .label {
    grid-area: label;
    font-family: "Lato", sans-serif;
    font-style: normal;
    font-weight: 600;
    font-size: 1rem;
    line-height: 1.5rem;
    display: flex;
    color: #26272B;
    margin-bottom: 0.5rem;
  }

  .autoHeight {
    height: auto;
  }

  .input-box {
    textarea {
      appearance: none;
      background: white;
      border: 1px solid #D1D1D6;
      border-radius: 0.50rem;
      box-sizing: border-box;
      color: #26272B;
      font-family: var(--font-nunito-sans);
      font-style: normal;
      font-weight: 400;
      font-size: 1rem;
      line-height: 24px;
      padding: 0.5em 0.75em;
      transition: all 0.25s;
      width: 100%;
      min-width: 100%;
      max-width: 100%;
      min-height: 10.5rem;
      resize: vertical;
      max-height: 348px;
      overflow: hidden;

      &:focus {
        border: 0.0625rem solid #B4000F;
        outline: none;
      }

      &:focus:not(.error) {
        border: 0.0625rem solid #B4000F;
        outline: none;
      }

      &::placeholder {
        color: #969492;
      }
    }

    .textarea::-webkit-scrollbar {
      display: none;
    }
  }

  .disabledLabel {
    color: #C9C7C5 !important;
  }

  .error {
    color: #B4000F;
  }

  .errorInput {
    color: #B4000F;
    border: 0.0625rem solid #B4000F !important;
  }

  .empty {
    color: #DDDBDA;
    font-size: 1em;
    font-family: 'Lato', sans-serif;
  }

  .message {
    grid-area: message;
    height: 24px;
    font-size: 1em;
    font-family: 'Lato', sans-serif;
  }

  .required {
    color: #B4000F;
    margin-left: 0.2rem;
  }