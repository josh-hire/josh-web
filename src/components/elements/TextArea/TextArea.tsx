import React, { useRef } from 'react';
import clsx from 'clsx';
import styles from './TextArea.module.scss';

interface Meta {
  dirty?: boolean;
  error?: string;
  touched?: boolean;
  submitError?: string;
}

interface TextAreaProps {
  className?: string;
  disabled?: boolean;
  label?: string;
  name?: string;
  meta?: Meta;
  icon?: React.ReactNode;
  inputClassName?: string;
  input?: React.TextareaHTMLAttributes<HTMLTextAreaElement>;
  required?: boolean;
  message?: string;
  onClickIcon?: () => void;
  onKeyUp?: (event: React.KeyboardEvent<HTMLTextAreaElement>) => void;
  insetIcon?: React.ReactNode;
  errorAutoHeight?: boolean;
  [key: string]: any;
}

const TextArea: React.FC<TextAreaProps> = ({
  className,
  disabled,
  label,
  name,
  meta = { dirty: false, touched: false },
  icon: Icon,
  inputClassName,
  input,
  required,
  message,
  onClickIcon,
  onKeyUp,
  insetIcon: InsetIcon,
  errorAutoHeight = false,
  ...props
}) => {
  const { dirty, error, touched, submitError } = meta;
  const textAreaRef = useRef<HTMLTextAreaElement | null>(null);

  const textAutoGrow = () => {
    const elm = textAreaRef.current;
    if (elm) {
      elm.style.height = '168px';
      if (!elm.value) {
        elm.style.height = '168px';
      } else {
        elm.style.height = `${elm.scrollHeight}px`;
      }
      if (elm.scrollHeight > 336) {
        elm.style.height = '336px';
        elm.style.overflowY = 'scroll';
      } else {
        elm.style.overflow = 'hidden';
      }
    }
  };

  return (
    <div
      className={clsx(
        className,
        styles.container,
        Icon && styles['container-icon'],
        InsetIcon && styles['container-inset-icon'],
      )}
      {...props}
    >
      <label className={clsx(styles.label, disabled && styles.disabledLabel)}>
        {label}
        {required && <span className={styles.required}>*</span>}
      </label>
      <div className={styles.icon}>
        {Icon && (
          <span className={styles.icon} onClick={onClickIcon}>
            {Icon}
          </span>
        )}
      </div>
      <div className={styles['input-box']}>
        <textarea
          autoComplete="off"
          rows={1}
          style={{ height: '168px' }}
          {...input}
          className={clsx(
            inputClassName,
            styles.input,
            !!input.value && styles.empty,
            (!!error || !!submitError) && (dirty || touched) && styles.errorInput,
          )}
          id={name}
          {...props}
          onInput={textAutoGrow}
          onKeyUp={onKeyUp}
          ref={textAreaRef}
        />
      </div>
      <div className={clsx(styles.message, errorAutoHeight && styles.autoHeight)}>
        {message && (
          <span className={styles.message}>
            <a>{message}</a>
          </span>
        )}
        {(!!error || !!submitError) && (dirty || touched) && (
          <small className={styles.error}>{error || submitError}</small>
        )}
      </div>
    </div>
  );
};

export default TextArea;
