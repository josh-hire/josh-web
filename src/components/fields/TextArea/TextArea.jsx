import { Textarea as Base } from '@mantine/core';
import PropTypes from 'prop-types';

export default function TextArea({
  autosize,
  disabled,
  icon,
  iconWidth,
  invalid,
  maxRows,
  minRows,
  radius,
  rightSection,
  rightSectionWidth,
  ...rest
}) {
  return (
    <Base
      autosize={autosize}
      disabled={disabled}
      icon={icon}
      iconWidth={iconWidth}
      invalid={invalid}
      maxRows={maxRows}
      minRows={minRows}
      radius={radius}
      rightSection={rightSection}
      rightSectionWidth={rightSectionWidth}
      {...rest}
    />
  );
}


TextArea.defaultProps = {
  autosize: false,
  disabled: false,
  icon: null,
  iconWidth: 36,
  invalid: false,
  maxRows: 0,
  minRows: 0,
  radius: 3,
  rightSection: null,
  rightSectionWidth: 36,
};

TextArea.propTypes = {
  autosize: PropTypes.bool,
  disabled: PropTypes.bool,
  icon: PropTypes.node,
  iconWidth: PropTypes.number,
  invalid: PropTypes.bool,
  maxRows: PropTypes.number,
  minRows: PropTypes.number,
  radius: PropTypes.number,
  rightSection: PropTypes.node,
  rightSectionWidth: PropTypes.number,
};
