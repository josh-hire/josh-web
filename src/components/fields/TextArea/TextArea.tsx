import { Textarea as Base, TextareaProps } from '@mantine/core';
import React from 'react';

interface TextAreaProps extends TextareaProps {
  icon?: React.ReactNode;
  iconWidth?: number;
  rightSection?: React.ReactNode;
  rightSectionWidth?: number;
}

const TextArea: React.FC<TextAreaProps> = ({
  autosize = false,
  disabled = false,
  icon = null,
  iconWidth = 36,
  maxRows = 0,
  minRows = 0,
  radius = 3,
  rightSection = null,
  rightSectionWidth = 36,
  ...rest
}) => {
  return (
    <Base
      autosize={autosize}
      disabled={disabled}
      icon={icon}
      iconWidth={iconWidth}
      maxRows={maxRows}
      minRows={minRows}
      radius={radius}
      rightSection={rightSection}
      rightSectionWidth={rightSectionWidth}
      {...rest}
    />
  );
};

export default TextArea;
