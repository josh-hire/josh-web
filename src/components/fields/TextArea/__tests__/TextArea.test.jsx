import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import TextArea from '../index';

describe('src/components/fields/TextArea', ()=>{
  test('TextArea',()=>{
    const { asFragment } = render(<TextArea />);
    expect(asFragment()).toMatchSnapshot();
  });
});
