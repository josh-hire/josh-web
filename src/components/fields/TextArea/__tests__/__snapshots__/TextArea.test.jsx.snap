// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fields/TextArea TextArea 1`] = `
<DocumentFragment>
  <div
    class="mantine-InputWrapper-root mantine-Textarea-root mantine-1m3pqry"
  >
    <div
      class="mantine-Input-wrapper mantine-Textarea-wrapper mantine-12sbrde"
    >
      <textarea
        aria-invalid="false"
        class="mantine-Input-input mantine-Textarea-input mantine-5mc5fy"
        id="mantine-r0"
      />
    </div>
  </div>
</DocumentFragment>
`;
